subprojects/docs/src/docs/userguide/developing-plugins/writing_in_kotlin.adoc

```diff
@@ -0,0 +1,329 @@
+
+TIP: Because it is provided by Gradle at runtime, `kotlin-stdlib` must be a `compileOnly` dependency.
+
+[[sec:kotlin_reflect]]
+=== kotlin-reflect
+
+[source,kotlin]
+----
+dependencies {
+    compileOnly("org.jetbrains.kotlin:kotlin-reflect:$version")
+}
+----
+
+Not all plugins need `kotlin-reflect` but if yours does, the situation is the same as for `kotlin-stdlib`. `kotlin-reflect` needs to be compatible with the https://docs.gradle.org/current/userguide/compatibility.html#kotlin[target Gradle version]
+
+TIP: Because it is provided by Gradle at runtime, `gradle-api` must be a `compileOnly` dependency.
+
+[[sec:others]]
+=== others (`implementation` or `api`)
+
+You can add other non-compileOnly dependencies like in other JVM projects:
+
+[source,kotlin]
+----
+dependencies {
+    // Kotlin dependencies
+    // ⚠️ make sure they do not depend on an incompatible version of kotlin-stdlib
+    implementation("com.squareup:kotlinpoet:1.12.0")
+
+    // Java dependencies
+    implementation("asm:asm:3.3.1")
+}
+----
+
+These dependencies are not provided by Gradle at runtime and should use the `implementation` configuration (or `api` if part of your plugin API). Because Gradle may load different plugins with conflicting dependencies in the same https://dev.to/autonomousapps/build-compile-run-a-crash-course-in-classpaths-f4g[classloader], your plugin users may experience crashes if symbols are incompatible or missing. To mitigate those, you can relocate your dependencies.
+
+[[sec:compiler_options]]
+== Different compiler configurations
+
+You can choose different options to either improve the developer experience of writing plugins and/or improve compatibility with Gradle and Java in general.
+
+[[sec:sam_with_receiver]]
+=== `sam-with-receiver` compiler plugin
+
+Gradle comes with the https://docs.gradle.org/current/javadoc/org/gradle/api/HasImplicitReceiver.html[@HasImplicitReceiver] annotation. When used in conjunction with Kotlin https://kotlinlang.org/docs/sam-with-receiver-plugin.html[sam-with-receiver compiler plugin], it can turn the parameter of https://kotlinlang.org/docs/fun-interfaces.html[Single Abstract Method (SAM)] interfaces like https://docs.gradle.org/current/javadoc/org/gradle/api/Action.html[Action] into implicit receiver making your code look more like a DSL:
+
+[source,kotlin]
+----
+// Without sam-with-receiver
+tasks.register("hello") { task ->
+    // You need to use the lambda parameter
+    task.doLast {
+        println("Hello World")
+    }
+}
+
+// With sam-with-receiver
+tasks.register("hello") {
+    // You can use the receiver
+    doLast {
+        println("Hello World")
+    }
+}
+----
+
+The latter looks closer to `build.gradle.kts` scripts and more concise but is more prone to scoping errors as well.
+
+
+[[sec:kotlin_assignment]]
+=== `kotlin-assignment` compiler plugin
```

Comment:
Is this worth mentioning yet? There's no instructions for how to use it or docs about it, and it seems generally unavailable.

Even when it's available though, is this a good API for plugin authors to use? Seems this should really be targeted solely for build scripts that are mostly using DSLs

Reply:
Good point. I put it for the sake of exhaustivity but not sure it's helpful to plugin authors. 

As a matter of fact, my preference goes to not using neither `sam-with-receiver` nor precompiled scripts plugins when developing custom plugins too. But there are lots of folks using `kotlin-dsl` to develop custom plugins so 🤷

Reply:
There's a lot of gotcha with using Kotlin for plugin writing, so some features may not be recommended for the task. The Gradle team should pitch in on this.